File: server/src/main/java/com/smartexpense/api/infrastructure/persistence/mapper/ExpenseMapper.java
package com.smartexpense.api.infrastructure.persistence.mapper;

import com.smartexpense.api.domain.model.Expense;
import com.smartexpense.api.infrastructure.persistence.entity.ExpenseEntity;
import com.smartexpense.api.infrastructure.persistence.entity.UserEntity;
import org.springframework.stereotype.Component;

@Component
public class ExpenseMapper {

    public Expense toDomain(ExpenseEntity entity) {
        if (entity == null) {
            return null;
        }
        return Expense.builder()
                .id(entity.getId())
                .userId(entity.getUser().getId())
                .amount(entity.getAmount())
                .category(entity.getCategory())
                .photoUrl(entity.getPhotoUrl())
                .caption(entity.getCaption())
                .expenseDate(entity.getExpenseDate())
                .createdAt(entity.getCreatedAt())
                .updatedAt(entity.getUpdatedAt())
                .build();
    }

    public ExpenseEntity toEntity(Expense domain) {
        if (domain == null) {
            return null;
        }
        UserEntity user = new UserEntity();
        user.setId(domain.getUserId());

        return ExpenseEntity.builder()
                .id(domain.getId())
                .user(user)
                .amount(domain.getAmount())
                .category(domain.getCategory())
                .photoUrl(domain.getPhotoUrl())
                .caption(domain.getCaption())
                .expenseDate(domain.getExpenseDate())
                .createdAt(domain.getCreatedAt())
                .updatedAt(domain.getUpdatedAt())
                .build();
    }
}

File: server/bin/main/com/smartexpense/api/infrastructure/persistence/mapper/ExpenseMapper.kt
package com.smartexpense.api.infrastructure.persistence.mapper

import com.smartexpense.api.domain.model.Expense
import com.smartexpense.api.infrastructure.persistence.entity.ExpenseEntity
import com.smartexpense.api.infrastructure.persistence.entity.UserEntity
import org.springframework.stereotype.Component

@Component
class ExpenseMapper {

    fun toDomain(entity: ExpenseEntity) = Expense(
        id = entity.id,
        userId = entity.user.id!!,
        amount = entity.amount,
        category = entity.category,
        photoUrl = entity.photoUrl,
        caption = entity.caption,
        expenseDate = entity.expenseDate,
        createdAt = entity.createdAt,
        updatedAt = entity.updatedAt
    )

    fun toEntity(domain: Expense): ExpenseEntity {
        // Create a proxy UserEntity with only the ID to satisfy the FK relationship
        val userRef = UserEntity(
            id = domain.userId,
            email = "",
            passwordHash = "",
            displayName = ""
        )
        return ExpenseEntity(
            id = domain.id,
            user = userRef,
            amount = domain.amount,
            category = domain.category,
            photoUrl = domain.photoUrl,
            caption = domain.caption,
            expenseDate = domain.expenseDate,
            createdAt = domain.createdAt,
            updatedAt = domain.updatedAt
        )
    }
}
